File: python/934. Shortest Bridge.py
class Solution(object):
    def shortestBridge(self, A):
        """
        :type A: List[List[int]]
        :rtype: int
        """
        m, n = len(A), len(A[0])

        x = y = 0
        for i in range(m):
            for j in range(n):
                if A[i][j] == 1:
                    x, y = i, j
                    break

        offset = ((1,0), (-1,0), (0,1), (0,-1))
        vis = [[0]*n for _ in range(m)]
        vis[x][y] = 1
        que = [(x, y)]
        def dfs(x, y, que):
            for dx, dy in offset:
                nx = x + dx
                ny = y + dy
                if 0<=nx<m and 0<=ny<n and A[nx][ny]==1 and not vis[nx][ny]:
                    que.append((nx, ny))
                    vis[nx][ny] = 1
                    dfs(nx, ny, que)

        dfs(x, y, que)
        # print que
        step = 0
        while que:
            sz = len(que)
            while sz:
                sz -= 1
                x, y = que.pop(0)
                if A[x][y]==1 and step > 0:
                    return step - 1
                for dx, dy in offset:
                    nx = x + dx
                    ny = y + dy
                    if 0<=nx<m and 0<=ny<n and not vis[nx][ny]:
                        que.append((nx, ny))
                        vis[nx][ny] = 1
            step += 1
        return step


# [[1,1,1,1,1],
# [1,0,0,0,1],
# [1,0,1,0,1],
# [1,0,0,0,1],
# [1,1,1,1,1]]

File: go/src/leetcode/934/934. Shortest Bridge.go
/**
@author ZhengHao Lou
Date    2022/10/25
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/shortest-bridge/
*/
var diff = [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

func shortestBridge(grid [][]int) int {
	m, n := len(grid), len(grid[0])
	var que [][]int
	var vis = make([][]bool, m)
	for i := range vis {
		vis[i] = make([]bool, n)
	}

	var dfs func(x, y int)
	dfs = func(x, y int) {
		for _, d := range diff {
			nx, ny := x+d[0], y+d[1]
			if nx >= 0 && nx < m && ny >= 0 && ny < n && !vis[nx][ny] && grid[nx][ny] == 1 {
				que = append(que, []int{nx, ny})
				vis[nx][ny] = true
				grid[nx][ny] = 2
				dfs(nx, ny)
			}
		}
	}

OUT:
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == 1 {
				vis[i][j] = true
				grid[i][j] = 2
				que = append(que, []int{i, j})
				dfs(i, j)
				break OUT
			}
		}
	}

	var bfs func() int
	bfs = func() int {
		for step := 0; len(que) != 0; step++ {
			for tmp := len(que); tmp != 0; tmp-- {
				x, y := que[0][0], que[0][1]
				que = que[1:]
				if grid[x][y] == 1 {
					return step - 1
				}
				for _, d := range diff {
					nx, ny := x+d[0], y+d[1]
					if nx >= 0 && nx < m && ny >= 0 && ny < n && !vis[nx][ny] {
						vis[nx][ny] = true
						que = append(que, []int{nx, ny})
					}
				}
			}
		}
		return -1
	}

	return bfs()
}

func main() {
	fmt.Println(shortestBridge([][]int{{0, 1}, {1, 0}}))
	fmt.Println(shortestBridge([][]int{{0, 1, 0}, {0, 0, 0}, {0, 0, 1}}))
}
